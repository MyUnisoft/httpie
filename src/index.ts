// Import Third-party Dependencies
import {
  Agent,
  ProxyAgent,
  fetch,
  setGlobalDispatcher,
  getGlobalDispatcher,
  Headers,
  HeadersInit,
  FormData,
  File,
  BodyInit,
  BodyMixin,
  MockAgent,
  mockErrors,
  MockPool,
  Interceptable,
  Client
} from "undici";

export * from "./request";
export * from "./stream";
export * from "./retry";
export * as policies from "./policies";
export { agents, computeURI, CustomHttpAgent } from "./agents";
export { DEFAULT_HEADER } from "./utils";

export {
  Agent,
  ProxyAgent,
  fetch,
  setGlobalDispatcher,
  getGlobalDispatcher,
  Headers,
  HeadersInit,
  FormData,
  File,
  BodyInit,
  BodyMixin,
  MockAgent,
  mockErrors,
  MockPool,
  Interceptable,
  Client
};
