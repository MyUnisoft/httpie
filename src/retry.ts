// Import Internal Dependencies
import Operation, { OperationResult } from "./class/Operation.class";
import { PolicyCallback, none } from "./policies";

/**
 * Those options are inspired by the retry package
 * @see https://www.npmjs.com/package/retry
 */
export interface RetryOptions {
  retries?: number;
  minTimeout?: number;
  maxTimeout?: number;
  unref?: boolean;
  factor?: number;
  forever?: boolean;
  signal?: AbortSignal | null;
}

export type RetryCallback<T> = (() => Promise<T>) | (() => T);

export async function retry<T>(
  callback: RetryCallback<T>, options: RetryOptions = {}, policy: PolicyCallback = none
): Promise<OperationResult<T>> {
  const op = new Operation<T>(options);

  do {
    try {
      const data = await callback();
      op.success(data);
    }
    catch (error: any) {
      if (policy(error)) {
        throw error;
      }

      await op.retry();
    }
  } while (op.continue);

  return op.toJSON();
}
