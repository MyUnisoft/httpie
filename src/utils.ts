/* eslint-disable no-redeclare */

// Import Node.js Dependencies
import { IncomingHttpHeaders } from "node:http";

// Import Internal Dependencies
import { RequestOptions, RequestResponse } from "./request";
import { HttpieError } from "./class/HttpieCommonError";
import { HttpieOnHttpError } from "./class/HttpieOnHttpError";

// CONSTANTS
const kDefaultUserAgent = "httpie";
const kDefaultEncodingCharset = "utf-8";
const kCharsetConversionTable = {
  "ISO-8859-1": "latin1"
};

export const DEFAULT_HEADER = { "user-agent": kDefaultUserAgent };

export function isAsyncIterable(value: any): boolean {
  return typeof value[Symbol.asyncIterator] === "function";
}

/**
 * @description Get a valid Node.js charset from the "content-type" http header.
 * @see https://nodejs.org/api/buffer.html#buffer_buffers_and_character_encodings
 */
export function getEncodingCharset(charset = kDefaultEncodingCharset): BufferEncoding {
  if (Buffer.isEncoding(charset)) {
    return charset as BufferEncoding;
  }

  return charset in kCharsetConversionTable ? kCharsetConversionTable[charset] : "utf-8";
}

/**
 * @description Create a default plain Object headers that will contains a Set of default values like:
 * - User-agent
 * - Authorization
 */
export function createHeaders(options: Partial<Pick<RequestOptions, "headers" | "authorization">>): IncomingHttpHeaders {
  const headers = Object.assign({ ...DEFAULT_HEADER }, options.headers ?? {});

  if (options.authorization) {
    headers.Authorization = createAuthorizationHeader(options.authorization);
  }

  return headers;
}

export function isHttpieError(error: unknown): error is HttpieError {
  return error instanceof HttpieError;
}

export function isHTTPError<T extends RequestResponse<any> = RequestResponse<any>>(
  error: unknown
): error is HttpieOnHttpError<T> {
  return error instanceof HttpieOnHttpError;
}

export function createBody(body: undefined): undefined;
export function createBody(body: any, headers?: IncomingHttpHeaders): string | Buffer;

/**
 * @description Generate a proper body for Undici Client. This method was mainly created to automatically manage JSON content.
 */
export function createBody(body: any, headers: IncomingHttpHeaders = {}): string | Buffer | undefined {
  if (typeof body === "undefined") {
    return void 0;
  }
  if (isAsyncIterable(body)) {
    return body;
  }

  let finalBody = body;
  if (body instanceof URLSearchParams) {
    headers["content-type"] = "application/x-www-form-urlencoded";
    finalBody = body.toString();
  }
  else if (typeof body === "object" && !Buffer.isBuffer(body)) {
    headers["content-type"] = "application/json";
    finalBody = JSON.stringify(body);
  }
  headers["content-length"] = String(Buffer.byteLength(finalBody));

  return finalBody;
}

/**
 * @description Helpers to generate a Basic or Bearer token for the HTTP Authorization header.
 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization
 */
export function createAuthorizationHeader(authorizationHeaderValue: string): string {
  const isBasicAuthToken = authorizationHeaderValue.includes(":");

  return isBasicAuthToken ?
    `Basic ${Buffer.from(authorizationHeaderValue).toString("base64")}` :
    `Bearer ${authorizationHeaderValue}`;
}

export const env = process.env;

export function getCurrentEnv() {
  const currentEnv: string = (env.NODE_ENV ?? "dev").toLowerCase();

  if (currentEnv.startsWith("prod")) {
    return "prod";
  }
  else if (currentEnv.startsWith("staging") || currentEnv.startsWith("preprod")) {
    return "preprod";
  }

  return "dev";
}
