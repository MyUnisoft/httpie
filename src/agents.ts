// Import Third-party Dependencies
import { Agent, ProxyAgent, MockAgent } from "undici";
import { LRUCache } from "lru-cache";

// Import Internal Dependencies
import { InlineCallbackAction, HttpMethod, WebDavMethod } from "./request";
import { getCurrentEnv } from "./utils";

// CONSTANTS
const kEnvName = getCurrentEnv();

/**
 * @see https://en.wikipedia.org/wiki/Page_replacement_algorithm
 */
export const URICache = new LRUCache<string | URL, computedUrlAndAgent>({
  max: 100,
  ttl: 1_000 * 60 * 120
});

export interface computedUrlAndAgent {
  url: URL;
  agent: Agent | ProxyAgent | MockAgent | null;
  limit?: InlineCallbackAction;
}

/**
 * These are agents specifically designed to work with MyUnisoft.
 */
export interface CustomHttpAgent {
  customPath: string;
  domains: Set<string>;
  agent: Agent | ProxyAgent | MockAgent;
  prod: string;
  preprod: string;
  dev: string;
  limit?: InlineCallbackAction;
}

export const agents: Set<CustomHttpAgent> = new Set();

/**
 * @description Detect if a given string URI is matching a given Agent custom path.
 *
 * @example
 * const URI = computeAgentPath("/windev/ws_monitoring", windev);
 * assert.strictEqual(URI, "https://ws-dev.myunisoft.fr/ws_monitoring");
 */
export function isAgentPathMatchingURI(uri: string, agent: CustomHttpAgent): URL | null {
  // Note: we want to match both '/path/xxx...' and 'path/xxx...'
  const localCustomPath = uri.charAt(0) === "/" ? `/${agent.customPath}` : agent.customPath;

  return uri.startsWith(localCustomPath) ?
    new URL(uri.slice(localCustomPath.length), agent[kEnvName]) :
    null;
}

/**
 * @description Compute a given string URI to the local list of agents.
 */
export function computeURIOnAllAgents(uri: string): computedUrlAndAgent {
  for (const agent of agents) {
    const url = isAgentPathMatchingURI(uri, agent);

    if (url !== null) {
      return { url, agent: agent.agent, limit: agent.limit };
    }
  }
  const url = new URL(uri);
  const agent = detectAgentFromURI(url);

  return {
    url, agent: agent?.agent ?? null, limit: agent?.limit
  };
}

/**
 * @description Seek correspondence with local agents through the URI hostname
 * @see https://nodejs.org/api/url.html#url_url_hostname
 *
 * @example
 * detectAgentFromURI("https://ws-dev.myunisoft.fr/ws_monitoring"); // windev agent
 * detectAgentFromURI("https://www.google.fr/"); // null
 */
export function detectAgentFromURI(uri: URL): CustomHttpAgent | null {
  const hostname = uri.hostname;

  for (const agent of agents) {
    if (agent.domains.has(hostname)) {
      return agent;
    }
  }

  return null;
}

/**
 * @description Compute a given URI (format string or WHATWG URL) and return a fully build URL and paired agent.
 * Under the hood it use a LRU cache
 */
export function computeURI(
  method: HttpMethod | WebDavMethod,
  uri: string | URL
): computedUrlAndAgent {
  const uriStr = method.toUpperCase() + uri.toString();
  if (URICache.has(uriStr)) {
    return URICache.get(uriStr)!;
  }

  let response: computedUrlAndAgent;
  if (typeof uri === "string") {
    response = computeURIOnAllAgents(uri);
  }
  else {
    const agent = detectAgentFromURI(uri);

    response = { url: uri, agent: agent?.agent ?? null, limit: agent?.limit };
  }
  URICache.set(uriStr, response);

  return response;
}
