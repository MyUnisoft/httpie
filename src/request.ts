// Import Node.js Dependencies
import { IncomingHttpHeaders } from "node:http";
import { URLSearchParams } from "node:url";

// Import Third-party Dependencies
import * as undici from "undici";
import { Result } from "@openally/result";
import status from "statuses";

// Import Internal Dependencies
import * as Utils from "./utils";
import { computeURI } from "./agents";
import { HttpieResponseHandler, ModeOfHttpieResponseHandler } from "./class/undiciResponseHandler";
import { HttpieOnHttpError } from "./class/HttpieOnHttpError";
import { HttpieError } from "./class/HttpieCommonError";
import { HttpieDecompressionError, HttpieFetchBodyError, HttpieParserError } from "./class/HttpieHandlerError";

export type WebDavMethod = "MKCOL" | "COPY" | "MOVE" | "LOCK" | "UNLOCK" | "PROPFIND" | "PROPPATCH";
export type HttpMethod = "GET" | "HEAD" | "POST" | "PUT" | "DELETE" | "CONNECT" | "OPTIONS" | "TRACE" | "PATCH" ;
export type InlineCallbackAction = <T>(fn: () => Promise<T>) => Promise<T>;

export type RequestError<T> =
  Partial<HttpieOnHttpError<RequestResponse<T>>> &
  Partial<HttpieDecompressionError> &
  Partial<HttpieFetchBodyError> &
  Partial<HttpieParserError> &
  HttpieError;

export interface RequestOptions {
  /** @default 0 */
  maxRedirections?: number;
  /** @default{ "user-agent": "httpie" } */
  headers?: IncomingHttpHeaders;
  querystring?: string | URLSearchParams;
  body?: any;
  authorization?: string;
  // Could be dynamically computed depending on the provided URI.
  agent?: undici.Agent | undici.ProxyAgent | undici.MockAgent;
  /** @description API limiter from a package like `p-ratelimit`. */
  limit?: InlineCallbackAction;
  /** @default "parse" */
  mode?: ModeOfHttpieResponseHandler;
  /** @default true */
  throwOnHttpError?: boolean;
}

export interface RequestResponse<T> {
  data: T;
  headers: IncomingHttpHeaders;
  statusMessage: string;
  statusCode: number;
}

/**
 * @description httpie "like" request wrapper that use new Node.js http client undici under the hood.
 * @see https://github.com/nodejs/undici
 *
 * @example
 * const { statusCode, data } = await request("GET", "https://ws.dev.myunisoft.tech/ws_monitoring");
 * console.log(statusCode, data); // 200 "true"
 */
export async function request<T>(
  method: HttpMethod | WebDavMethod,
  uri: string | URL,
  options: RequestOptions = {}
): Promise<RequestResponse<T>> {
  const { maxRedirections = 0 } = options;

  const computedURI = computeURI(method, uri);
  if (typeof options.querystring !== "undefined") {
    const qs = typeof options.querystring === "string" ? new URLSearchParams(options.querystring) : options.querystring;
    for (const [key, value] of qs.entries()) {
      computedURI.url.searchParams.set(key, value);
    }
  }

  const limit = options.limit ?? computedURI.limit ?? null;
  const dispatcher = options.agent ?? computedURI.agent ?? void 0;

  const headers = Utils.createHeaders({ headers: options.headers, authorization: options.authorization });
  const body = Utils.createBody(options.body, headers);

  const requestOptions = { method: method as HttpMethod, headers, body, dispatcher, maxRedirections };
  const requestResponse = limit === null ?
    await undici.request(computedURI.url, requestOptions) :
    await limit(() => undici.request(computedURI.url, requestOptions));

  const statusCode = requestResponse.statusCode;
  const responseHandler = new HttpieResponseHandler(requestResponse);

  let data;
  if (options.mode === "parse" || !options.mode) {
    data = await responseHandler.getData<T>("parse");
  }
  else {
    data = await responseHandler.getData(options.mode);
  }

  const RequestResponse = {
    headers: requestResponse.headers,
    statusMessage: status.message[requestResponse.statusCode]!,
    statusCode,
    data
  };

  const shouldThrowOnHttpError = options.throwOnHttpError ?? true;
  if (shouldThrowOnHttpError && statusCode >= 400) {
    throw new HttpieOnHttpError(RequestResponse);
  }

  return RequestResponse;
}

export async function safeRequest<T, E>(
  method: HttpMethod | WebDavMethod,
  uri: string | URL,
  options: RequestOptions = {}
): Promise<Result<RequestResponse<T>, RequestError<E>>> {
  return Result.wrapAsync<RequestResponse<T>, RequestError<E>>(
    () => request(method, uri, options)
  );
}

export type RequestCallback = <T>(
  uri: string | URL, options?: RequestOptions
) => Promise<RequestResponse<T>>;
export type SafeRequestCallback = <T, E>(
  uri: string | URL, options?: RequestOptions
) => Promise<Result<RequestResponse<T>, RequestError<E>>>;

export const get = request.bind(null, "GET") as RequestCallback;
export const post = request.bind(null, "POST") as RequestCallback;
export const put = request.bind(null, "PUT") as RequestCallback;
export const del = request.bind(null, "DELETE") as RequestCallback;
export const patch = request.bind(null, "PATCH") as RequestCallback;

export const safeGet = safeRequest.bind(null, "GET") as SafeRequestCallback;
export const safePost = safeRequest.bind(null, "POST") as SafeRequestCallback;
export const safePut = safeRequest.bind(null, "PUT") as SafeRequestCallback;
export const safeDel = safeRequest.bind(null, "DELETE") as SafeRequestCallback;
export const safePatch = safeRequest.bind(null, "PATCH") as SafeRequestCallback;
