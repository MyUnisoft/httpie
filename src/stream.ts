// Import Node.js Dependencies
import { Duplex, Writable } from "stream";

// Import Third-party Dependencies
import * as undici from "undici";

// Import Internal Dependencies
import { ReqOptions, HttpMethod, WebDavMethod } from "./request";
import { computeURI } from "./agents";
import * as Utils from "./utils";

export type StreamOptions = Omit<ReqOptions, "limit">;

export function pipeline(
  method: HttpMethod | WebDavMethod,
  uri: string | URL,
  options: StreamOptions = {}
): Duplex {
  const { maxRedirections = 0 } = options;

  const computedURI = computeURI(uri);
  if (typeof options.querystring !== "undefined") {
    const qs = typeof options.querystring === "string" ? new URLSearchParams(options.querystring) : options.querystring;
    for (const [key, value] of qs.entries()) {
      computedURI.url.searchParams.set(key, value);
    }
  }

  const dispatcher = options.agent ?? computedURI.agent ?? void 0;
  const headers = Utils.createHeaders({ headers: options.headers, authorization: options.authorization });
  const body = Utils.createBody(options.body, headers);

  return undici.pipeline(computedURI.url, {
    method: method as HttpMethod, headers, body, dispatcher, maxRedirections
  }, ({ body }) => body);
}

export type WritableStreamCallback = (writable: Writable) => Promise<undici.Dispatcher.StreamData>;

export function stream(
  method: HttpMethod | WebDavMethod,
  uri: string | URL,
  options: StreamOptions = {}
): WritableStreamCallback {
  const { maxRedirections = 0 } = options;
  const computedURI = computeURI(uri);

  const dispatcher = options.agent ?? computedURI.agent ?? void 0;
  const headers = Utils.createHeaders({ headers: options.headers, authorization: options.authorization });
  const body = Utils.createBody(options.body, headers);

  return (writable: Writable) => undici
    .stream(
      computedURI.url,
      { method: method as HttpMethod, headers, body, dispatcher, maxRedirections },
      () => writable
    );
}
