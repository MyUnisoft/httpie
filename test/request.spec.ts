// Import Third-party Dependencies
import { FastifyInstance } from "fastify";

// Import Internal Dependencies
import { get, post, put, patch, del, safeGet } from "../src/index";

// Helpers and mock
import { createServer } from "./server/index";
import { windev } from "./helpers";

let httpServer: FastifyInstance;
beforeAll(async() => {
  httpServer = await createServer();
});

afterAll(async() => {
  await httpServer.close();
});

describe("http.get", () => {
  it("should GET uptime from local fastify server", async() => {
    const { data } = await get<{ uptime: number }>("/local/");

    expect("uptime" in data).toStrictEqual(true);
    expect(typeof data.uptime).toStrictEqual("number");
  });

  it("should GET query parameters provided to fastify", async() => {
    const { data } = await get<{ name: string }>("/local/qs", {
      querystring: new URLSearchParams({
        name: "foobar"
      })
    });

    expect("name" in data).toStrictEqual(true);
    expect(data.name).toStrictEqual("foobar");
  });

  it("should GET uptime by following an HTTP redirection from local fastify server", async() => {
    const { data } = await get<{ uptime: number }>("/local/redirect", { maxRedirections: 1 });

    expect("uptime" in data).toStrictEqual(true);
    expect(typeof data.uptime).toStrictEqual("number");
  });

  it("should GET uptime through a limit function handler from local fastify server", async() => {
    let executed = false;
    // eslint-disable-next-line func-style
    const limit = (callback) => {
      executed = true;

      return callback();
    };
    const { data } = await get<{ uptime: number }>("/local/", { limit });

    expect("uptime" in data).toStrictEqual(true);
    expect(typeof data.uptime).toStrictEqual("number");
    expect(executed).toStrictEqual(true);
  });

  it("should GET response from windev ws-monitoring endpoint (without Agent)", async() => {
    const { data } = await get<string>("/windev/ws_monitoring");

    expect(data).toStrictEqual(true);
  });

  it("should GET response from windev ws-monitoring endpoint (with Agent)", async() => {
    const { data } = await get<string>("/windev/ws_monitoring", {
      agent: windev.agent
    });

    expect(data).toStrictEqual(true);
  });

  it("should GET json response from node.js health endpoint", async() => {
    const { data } = await get<any>("https://app.dev.myunisoft.tech/api/v1/ipa/healthz");

    expect(Object.keys(data).sort()).toMatchObject([
      "status", "version", "description", "checks"
    ].sort());
  });

  it("should throw a 404 Not Found error because the path is not known", async() => {
    expect.assertions(4);

    try {
      await get<string>("/windev/hlkezcjcke");
    }
    catch (error) {
      expect(error.name).toStrictEqual("HttpieOnHttpError");
      expect(error.statusCode).toStrictEqual(404);
      expect(error.statusMessage).toStrictEqual("Not Found");
      expect(error.data).toMatchSnapshot();
    }
  });

  it("should throw a 'HttpieParserError' with jsonError endpoint from local fastify server", async() => {
    expect.assertions(4);

    const expectedPayload = "{ 'foo': bar }";
    try {
      await get<string>("/local/jsonError");
    }
    catch (error) {
      expect(error.name).toStrictEqual("ResponseParsingError");
      expect(error.reason.name).toStrictEqual("SyntaxError");
      expect(error.text).toStrictEqual(expectedPayload);
      expect(error.buffer).toStrictEqual(Buffer.from(expectedPayload));
    }
  });
});

/**
 * @see https://jsonplaceholder.typicode.com/guide/
 */
describe("http.post", () => {
  it("should POST data on jsonplaceholder API", async() => {
    const body = {
      title: "foo",
      body: "bar",
      userId: 1
    };

    const { data } = await post<typeof body & { userId: number }>("https://jsonplaceholder.typicode.com/posts", { body });
    expect(typeof data.userId).toStrictEqual("number");
    expect(data).toMatchObject(body);
  });
});


describe("http.put", () => {
  it("should PUT data on jsonplaceholder API", async() => {
    const body = {
      id: 1,
      title: "foo",
      body: "bar",
      userId: 1
    };

    const { data } = await put<typeof body & { userId: number }>("https://jsonplaceholder.typicode.com/posts/1", { body });
    expect(data).toEqual(body);
  });
});

describe("http.patch", () => {
  it("should PATCH data on jsonplaceholder API", async() => {
    const body = {
      id: 1,
      title: "foo",
      userId: 1
    };

    const { data } = await patch<typeof body & { userId: number }>("https://jsonplaceholder.typicode.com/posts/1", {
      body: { title: "foo" }
    });
    expect(data).toMatchObject(body);
  });
});

describe("http.del", () => {
  it("should DELETE data on jsonplaceholder API", async() => {
    const { statusCode } = await del<any>("https://jsonplaceholder.typicode.com/posts/1", {
      body: { title: "foo" }
    });
    expect(statusCode).toStrictEqual(200);
  });
});

describe("http.safeGet", () => {
  it("should GET uptime from local fastify server", async() => {
    const result = await safeGet<{ uptime: number }, any>("/local/");

    expect(result.ok).toStrictEqual(true);
    const { data } = result.unwrap();
    expect("uptime" in data).toStrictEqual(true);
    expect(typeof data.uptime).toStrictEqual("number");
  });

  it("should throw a 404 Not Found error because the path is not known", async() => {
    const result = await safeGet<string, any>("/windev/hlkezcjcke");
    expect(result.err).toStrictEqual(true);

    if (result.err) {
      const error = result.val;
      console.log(error);

      expect(error.name).toStrictEqual("HttpieOnHttpError");
      expect(error.statusCode).toStrictEqual(404);
      expect(error.statusMessage).toStrictEqual("Not Found");
      expect(error.data).toMatchSnapshot();
    }
  });
});
