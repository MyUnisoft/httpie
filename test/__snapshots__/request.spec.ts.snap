// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`http.get should throw a 404 Not Found error because the path is not known 1`] = `
"<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>404 Not Found</title>
</head><body>
<h1>Not Found</h1>
<p>The requested URL was not found on this server.</p>
<hr>
<address>Apache/2.4.59 (Debian) Server at ws-dev.myunisoft.fr Port 443</address>
</body></html>
"
`;

exports[`http.safeGet should throw a 404 Not Found error because the path is not known 1`] = `
"<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>404 Not Found</title>
</head><body>
<h1>Not Found</h1>
<p>The requested URL was not found on this server.</p>
<hr>
<address>Apache/2.4.59 (Debian) Server at ws-dev.myunisoft.fr Port 443</address>
</body></html>
"
`;
